### PR TITLE
User Limits

### PR DESCRIPTION
**Pull request by shrinathprabhu**
Resolves [AR-2249](https://team-1624093970686.atlassian.net/browse/AR-2249).

## Changes
- Added userlimits section
- Added default limits
- Added logic for limited and unlimited user limit
- Added all the store related changes

## Screenshots
Screenshot from 2022-05-10 21-46-23

## Checklist

- [x] The branch name follows the format: `developer/AR-XXX-issue-name`.
- [x] The changes have been tested locally.

### CODE REVIEW

File: src/utils/constants.ts [hunk 1]
```diff
@@ -73,6 +75,42 @@ const CONFIGURE_TABS: ConfigureTab[] = [
   { type: 'wallet', label: 'Wallet' },
 ]
 
+type UserLimit = {
+  isUnlimited: boolean
+  value?: number
+  unit?: 'MB' | 'GB'
+}
```
Comment (ajithranka) [inline src/utils/constants.ts:82]: We can avoid impossible state by using a type union here.

```
type UserLimit = 
  | { isUnlimited: true }
  | { isUnlimited: false, value: number, unit: 'MB' | 'GB' }
```

File: src/utils/constants.ts [hunk 2]
```diff
@@ -73,6 +75,42 @@ const CONFIGURE_TABS: ConfigureTab[] = [
...
+  isUnlimited: false,
+  value: 2,
+  unit: 'MB',
+}
+
+type UserLimitType = 'Unlimited' | 'Limited'
+
+const storageValues = ['MB', 'GB']
+
+type BandwidthLimitUnits = {
```
Comment (ajithranka) [inline src/utils/constants.ts:98]: Use singular names for types.

```suggestion
type BandwidthLimitUnit = {
```

File: src/utils/constants.ts [hunk 3]
```diff
@@ -73,6 +75,42 @@ const CONFIGURE_TABS: ConfigureTab[] = [
...
+  isUnlimited: true,
+}
+
+const defaultUserLimit: UserLimit = {
+  isUnlimited: false,
+  value: 2,
+  unit: 'MB',
+}
+
+type UserLimitType = 'Unlimited' | 'Limited'
```
Comment (ajithranka) [inline src/utils/constants.ts:94]: Avoid the `Type` post-fix when naming types.

```suggestion
type UserLimit = 'Unlimited' | 'Limited'
```

**Review by ajithranka — APPROVED** [review]

File: src/components/app-configure/store/UserLimits.vue [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+</script>
+
+<template>
+  <section name="user-limits">
+    <SettingCard>
+      <template #title>User Limits</template>
+      <template #description
+        >If you would like to limit the bandwidth or storage for each of your
+        users, please specify the same below. <a>LEARN MORE</a></template
+      >
```
Comment (karthik-durai) [inline src/components/app-configure/store/UserLimits.vue:80]: looks like formatting issue, can you check @shrinathprabhu

File: src/components/app-configure/store/UserLimits.vue [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+@media screen and (min-width: 1024px) {
+  .limits-input-container > * {
+    flex: 0 0 calc(50% - 1rem);
+  }
+}
+</style>
+
+<style>
```
Comment (karthik-durai) [inline src/components/app-configure/store/UserLimits.vue:217]: why not move this to scoped style, I guess `!important` is not needed when moved.

File: src/components/lib/VDropdown/VDropdown.vue [hunk 6]
```diff
@@ -11,7 +11,7 @@ const props = defineProps({
     default: '',
   },
   modelValue: {
-    type: [String, Object],
+    type: [String, Object, Boolean],
```
Comment (karthik-durai) [inline src/components/lib/VDropdown/VDropdown.vue:14]: why `Object` type?

File: src/components/lib/VTextField/VTextField.vue [hunk 7]
```diff
@@ -195,6 +205,11 @@ input::placeholder {
   color: var(--text-grey);
 }
 
+.form-group[data-disabled='true'],
+input:disabled {
```
Comment (karthik-durai) [inline src/components/lib/VTextField/VTextField.vue:209]: I recommend using `class` names for styling unless it is a global style.

you thoughts @ajithranka

Comment (ajithranka) [reply src/components/lib/VTextField/VTextField.vue:209]: Once we make clean up the CSS, we can enforce this with a lint rule. It's probably okay for now. :-)

File: src/components/app-configure/store/UserLimits.vue (hunk 4 above)
Comment (ajithranka) [reply src/components/app-configure/store/UserLimits.vue:80]: This is controlled by the `htmlWhitespaceSensitivity` config in Prettier. The rule helps avoid leading and trailing spaces in inline elements which get rendered on the page. See this [description](https://prettier.io/blog/2018/11/07/1.15.0.html#whitespace-sensitive-formatting).

File: src/components/lib/VDropdown/VDropdown.vue (hunk 6 above)
Comment (shrinathprabhu) [reply src/components/lib/VDropdown/VDropdown.vue:14]: We also have cases where we send `{label: 'Dark Mode', value: 'dark'}`. This is tend to improve when we start creating component library

File: src/components/app-configure/store/UserLimits.vue (hunk 5 above)
Comment (shrinathprabhu) [reply src/components/app-configure/store/UserLimits.vue:217]: Actually it won't work in scoped style. Dropdown trigger is not part of this component but is part of Dropdown component and to style that we'll either need to style it in Dropdown component (where we can use scoped style without `!important` flag) or we'll need to do it globally or as an unscoped internal style with `!important` flag to override the defaults.

File: src/components/app-configure/store/UserLimits.vue [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+} from '@/store/configure.store'
+import {
+  userLimitOptions,
+  storageValues,
+  bandwidthUnits,
+  type BandwidthLimitUnit,
+} from '@/utils/constants'
+
+type UserLimitKind = 'Unlimited' | 'Limited'
+type UserLimit = {
```
Comment (ajithranka) [inline src/components/app-configure/store/UserLimits.vue:22]: Is this the same as `UserLimitState`?

**Review by ajithranka — APPROVED** [review]

**Review by karthik-durai — APPROVED** [review]

**Pull request merged**